### PR TITLE
Enhancement/eks adding service ipv4 range

### LINKED ISSUE #15505: EKS Support Service IPv4 range
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Add support in `aws_eks_cluster` for the new Service IPv4 range feature of EKS clusters: https://aws.amazon.com/about-aws/whats-new/2020/10/amazon-eks-supports-configurable-kubernetes-service-ip-address-range/

### New or Affected Resource(s)

* aws_eks_cluster

### Potential Terraform Configuration

```hcl
resource "aws_eks_cluster" "cluster" {
  name     = var.cluster-name
  serviceIpv4Cidr = "10.0.0.0/16" // something like this is what I would imagine
}
```

### References

I'm not aware of any other open issues for this feature.

### PR DESCRIPTION
**Pull request by philnichol**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #15505

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
Added service_ipv4_cidr input to aws_eks_cluster.
```

Output from acceptance testing:

```
$ TF_ACC=1 go test ./aws -v -count 1 -parallel 4 -run=TestAccAWSEksCluster_ -timeout 120m
=== RUN   TestAccAWSEksCluster_basic
=== PAUSE TestAccAWSEksCluster_basic
=== RUN   TestAccAWSEksCluster_EncryptionConfig
=== PAUSE TestAccAWSEksCluster_EncryptionConfig
=== RUN   TestAccAWSEksCluster_Version
=== PAUSE TestAccAWSEksCluster_Version
=== RUN   TestAccAWSEksCluster_Logging
=== PAUSE TestAccAWSEksCluster_Logging
=== RUN   TestAccAWSEksCluster_Tags
=== PAUSE TestAccAWSEksCluster_Tags
=== RUN   TestAccAWSEksCluster_VpcConfig_SecurityGroupIds
=== PAUSE TestAccAWSEksCluster_VpcConfig_SecurityGroupIds
=== RUN   TestAccAWSEksCluster_VpcConfig_EndpointPrivateAccess
=== PAUSE TestAccAWSEksCluster_VpcConfig_EndpointPrivateAccess
=== RUN   TestAccAWSEksCluster_VpcConfig_EndpointPublicAccess
=== PAUSE TestAccAWSEksCluster_VpcConfig_EndpointPublicAccess
=== RUN   TestAccAWSEksCluster_VpcConfig_PublicAccessCidrs
=== PAUSE TestAccAWSEksCluster_VpcConfig_PublicAccessCidrs
=== CONT  TestAccAWSEksCluster_basic
=== CONT  TestAccAWSEksCluster_VpcConfig_SecurityGroupIds
=== CONT  TestAccAWSEksCluster_Logging
=== CONT  TestAccAWSEksCluster_Version
--- PASS: TestAccAWSEksCluster_Logging (944.83s)
=== CONT  TestAccAWSEksCluster_VpcConfig_PublicAccessCidrs
--- PASS: TestAccAWSEksCluster_VpcConfig_SecurityGroupIds (957.92s)
=== CONT  TestAccAWSEksCluster_EncryptionConfig
--- PASS: TestAccAWSEksCluster_basic (992.14s)
=== CONT  TestAccAWSEksCluster_VpcConfig_EndpointPublicAccess
--- PASS: TestAccAWSEksCluster_EncryptionConfig (907.86s)
=== CONT  TestAccAWSEksCluster_VpcConfig_EndpointPrivateAccess
--- PASS: TestAccAWSEksCluster_VpcConfig_PublicAccessCidrs (1211.42s)
=== CONT  TestAccAWSEksCluster_Tags
--- PASS: TestAccAWSEksCluster_Version (2577.52s)
--- PASS: TestAccAWSEksCluster_VpcConfig_EndpointPublicAccess (1640.84s)
--- PASS: TestAccAWSEksCluster_Tags (840.69s)
--- PASS: TestAccAWSEksCluster_VpcConfig_EndpointPrivateAccess (2191.08s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	4058.874s
```
Adds a kubernetes_network_configuration block, currently the only field in the block is service_ipv4_cidr, which sets the IP range allocated to k8s services within a cluster

### CODE REVIEW

**Comment by philnichol** [comment]
reran acceptance test (with newly added acceptance test and latest commit that sets minItems on list:
```
$ TF_ACC=1 go test ./aws -v -count 1 -parallel 4 -run=TestAccAWSEksCluster_ -timeout 120m
=== RUN   TestAccAWSEksCluster_basic
=== PAUSE TestAccAWSEksCluster_basic
=== RUN   TestAccAWSEksCluster_EncryptionConfig
=== PAUSE TestAccAWSEksCluster_EncryptionConfig
=== RUN   TestAccAWSEksCluster_Version
=== PAUSE TestAccAWSEksCluster_Version
=== RUN   TestAccAWSEksCluster_Logging
=== PAUSE TestAccAWSEksCluster_Logging
=== RUN   TestAccAWSEksCluster_Tags
=== PAUSE TestAccAWSEksCluster_Tags
=== RUN   TestAccAWSEksCluster_VpcConfig_SecurityGroupIds
=== PAUSE TestAccAWSEksCluster_VpcConfig_SecurityGroupIds
=== RUN   TestAccAWSEksCluster_VpcConfig_EndpointPrivateAccess
=== PAUSE TestAccAWSEksCluster_VpcConfig_EndpointPrivateAccess
=== RUN   TestAccAWSEksCluster_VpcConfig_EndpointPublicAccess
=== PAUSE TestAccAWSEksCluster_VpcConfig_EndpointPublicAccess
=== RUN   TestAccAWSEksCluster_VpcConfig_PublicAccessCidrs
=== PAUSE TestAccAWSEksCluster_VpcConfig_PublicAccessCidrs
=== RUN   TestAccAWSEksCluster_NetworkConfig_ServiceIpv4Cidr
=== PAUSE TestAccAWSEksCluster_NetworkConfig_ServiceIpv4Cidr
=== CONT  TestAccAWSEksCluster_basic
=== CONT  TestAccAWSEksCluster_VpcConfig_EndpointPrivateAccess
=== CONT  TestAccAWSEksCluster_NetworkConfig_ServiceIpv4Cidr
=== CONT  TestAccAWSEksCluster_VpcConfig_PublicAccessCidrs
--- PASS: TestAccAWSEksCluster_basic (841.95s)
=== CONT  TestAccAWSEksCluster_VpcConfig_EndpointPublicAccess
--- PASS: TestAccAWSEksCluster_NetworkConfig_ServiceIpv4Cidr (935.96s)
=== CONT  TestAccAWSEksCluster_Logging
--- PASS: TestAccAWSEksCluster_VpcConfig_PublicAccessCidrs (1052.45s)
=== CONT  TestAccAWSEksCluster_VpcConfig_SecurityGroupIds
--- PASS: TestAccAWSEksCluster_Logging (880.39s)
=== CONT  TestAccAWSEksCluster_Tags
--- PASS: TestAccAWSEksCluster_VpcConfig_SecurityGroupIds (940.63s)
=== CONT  TestAccAWSEksCluster_Version
--- PASS: TestAccAWSEksCluster_VpcConfig_EndpointPrivateAccess (2220.05s)
=== CONT  TestAccAWSEksCluster_EncryptionConfig
--- PASS: TestAccAWSEksCluster_VpcConfig_EndpointPublicAccess (1669.62s)
--- PASS: TestAccAWSEksCluster_Tags (905.21s)
--- PASS: TestAccAWSEksCluster_EncryptionConfig (875.73s)
--- PASS: TestAccAWSEksCluster_Version (2711.32s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	4706.341s
```

**Comment by etoews** [comment]
@anGie44 We're in need of this functionality too. The PR is looking good so far. Does anything else need to happen before it can be merged and released? Thanks.

File: aws/resource_aws_eks_cluster_test.go
Comment (anGie44) [inline aws/resource_aws_eks_cluster_test.go:427]: if you don't mind, can you add a test for the case where the empty config block (`kubernetes_network_config {}`) is provided to test the case where AWS supplies the IP?

File: aws/resource_aws_eks_cluster_test.go
Comment (anGie44) [inline aws/resource_aws_eks_cluster_test.go:839]: looks great overall, just a small nitpick to use the helper func `composeConfig` to concatenate the base config here 
```suggestion
	return composeConfig(testAccAWSEksClusterConfig_Base(rName), fmt.Sprintf(`
resource "aws_eks_cluster" "test" {
  name     = %q
  role_arn = aws_iam_role.test.arn

  vpc_config {
    subnet_ids = aws_subnet.test[*].id
  }

  kubernetes_network_config {
    service_ipv4_cidr = %s
  }

  depends_on = [aws_iam_role_policy_attachment.test-AmazonEKSClusterPolicy]
}
`, rName, serviceIpv4Cidr)
}
```

File: aws/resource_aws_eks_cluster_test.go
Comment (anGie44) [inline aws/resource_aws_eks_cluster_test.go:459]: since we want to ensure no further Op is needed when the IP is unchanged, we can use a `PlanOnly` check in place here
```suggestion
			{
				Config:   testAccAWSEksClusterConfig_NetworkConfig_ServiceIpv4Cidr(rName, `"10.10.0.0/16"`),
				PlanOnly: true,
				ExpectNonEmptyPlan: false,
			},
```

Comment (anGie44) [reply aws/resource_aws_eks_cluster_test.go:459]: it would also be great to add a test to verify that the resource is recreated in cases where the IP changes if time permits :)

**Review by anGie44 — CHANGES_REQUESTED** [review]
Hi @philnichol, thank you for this PR! Off to a really great start 👍 A couple comments and suggestions to help address the current acctest failure and request for test coverage. If you have any questions or concerns, please reach out!

File: aws/resource_aws_eks_cluster.go
Comment (anGie44) [reply aws/resource_aws_eks_cluster.go:366]: test output atm:
```
 TestAccAWSEksCluster_NetworkConfig_ServiceIpv4Cidr: resource_aws_eks_cluster_test.go:433: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.
        
        (map[string]string) {
        }
        
        
        (map[string]string) (len=3) {
         (string) (len=27) "kubernetes_network_config.#": (string) (len=1) "1",
         (string) (len=29) "kubernetes_network_config.0.%": (string) (len=1) "1",
```

File: aws/resource_aws_eks_cluster_test.go
Comment (philnichol) [reply aws/resource_aws_eks_cluster_test.go:429]: Good shout, have added this to the _Basic test

**Comment by philnichol** [comment]
@anGie44  thanks for the feedback! All great points! I think I've addressed your comments but let me know if I've missed anything.
Here's the updated output from the acceptance test:
```
$ TF_ACC=1 go test ./aws -v -count 1 -parallel 4 -run=TestAccAWSEksCluster_ -timeout 120m
=== RUN   TestAccAWSEksCluster_basic
=== PAUSE TestAccAWSEksCluster_basic
=== RUN   TestAccAWSEksCluster_EncryptionConfig
=== PAUSE TestAccAWSEksCluster_EncryptionConfig
=== RUN   TestAccAWSEksCluster_Version
=== PAUSE TestAccAWSEksCluster_Version
=== RUN   TestAccAWSEksCluster_Logging
=== PAUSE TestAccAWSEksCluster_Logging
=== RUN   TestAccAWSEksCluster_Tags
=== PAUSE TestAccAWSEksCluster_Tags
=== RUN   TestAccAWSEksCluster_VpcConfig_SecurityGroupIds
=== PAUSE TestAccAWSEksCluster_VpcConfig_SecurityGroupIds
=== RUN   TestAccAWSEksCluster_VpcConfig_EndpointPrivateAccess
=== PAUSE TestAccAWSEksCluster_VpcConfig_EndpointPrivateAccess
=== RUN   TestAccAWSEksCluster_VpcConfig_EndpointPublicAccess
=== PAUSE TestAccAWSEksCluster_VpcConfig_EndpointPublicAccess
=== RUN   TestAccAWSEksCluster_VpcConfig_PublicAccessCidrs
=== PAUSE TestAccAWSEksCluster_VpcConfig_PublicAccessCidrs
=== RUN   TestAccAWSEksCluster_NetworkConfig_ServiceIpv4Cidr
=== PAUSE TestAccAWSEksCluster_NetworkConfig_ServiceIpv4Cidr
=== CONT  TestAccAWSEksCluster_basic
=== CONT  TestAccAWSEksCluster_VpcConfig_EndpointPrivateAccess
=== CONT  TestAccAWSEksCluster_NetworkConfig_ServiceIpv4Cidr
=== CONT  TestAccAWSEksCluster_VpcConfig_PublicAccessCidrs
--- PASS: TestAccAWSEksCluster_basic (810.21s)
=== CONT  TestAccAWSEksCluster_Logging
--- PASS: TestAccAWSEksCluster_VpcConfig_PublicAccessCidrs (1486.56s)
=== CONT  TestAccAWSEksCluster_VpcConfig_SecurityGroupIds
--- PASS: TestAccAWSEksCluster_Logging (1257.63s)
=== CONT  TestAccAWSEksCluster_VpcConfig_EndpointPublicAccess
--- PASS: TestAccAWSEksCluster_NetworkConfig_ServiceIpv4Cidr (2302.95s)
=== CONT  TestAccAWSEksCluster_Tags
--- PASS: TestAccAWSEksCluster_VpcConfig_SecurityGroupIds (900.22s)
=== CONT  TestAccAWSEksCluster_Version
--- PASS: TestAccAWSEksCluster_VpcConfig_EndpointPrivateAccess (2718.91s)
=== CONT  TestAccAWSEksCluster_EncryptionConfig
--- PASS: TestAccAWSEksCluster_Tags (934.24s)
--- PASS: TestAccAWSEksCluster_EncryptionConfig (737.58s)
--- PASS: TestAccAWSEksCluster_VpcConfig_EndpointPublicAccess (1723.76s)
--- PASS: TestAccAWSEksCluster_Version (2832.07s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       5218.886s
```

**Comment by etoews** [comment]
@anGie44 Thanks a lot for that initial review. Everything is looking good here. 

Do you have some time to give this another pass?

**Comment by jasonyihk** [comment]
we also in need for this functionality, please release asap. tks

**Comment by shrutilamba** [comment]
We need this functionality as well. How long will this take to be released?

**Review by anGie44 — CHANGES_REQUESTED** [review]
Hi @philnichol (and those following this feature 😄 )-- apologies for the delay here ! a couple final notes before we can merge this in 👍

File: website/docs/r/eks_cluster.html.markdown
Comment (anGie44) [inline website/docs/r/eks_cluster.html.markdown:159]: Let's also a note here that if this block is removed from the configuration though, its essentially a no-op in the prior state so we can't do drift detection unless it's configured

File: aws/resource_aws_eks_cluster_test.go
Comment (philnichol) [reply aws/resource_aws_eks_cluster_test.go]: Good spot!

File: aws/resource_aws_eks_cluster.go
Comment (philnichol) [reply aws/resource_aws_eks_cluster.go:118]: Hi @anGie44 , no worries at all. Happy to have a crack at the data_source, but if it takes a while I might well ask for your help since I don't want to hold people up :smile:. Before making any changes I seem to be getting the following error on all datasource tests, but can't see what I'm doing wrong, any ideas? I get the same (resourceNotFound) regardless of which datasource test I run it seems.
```
TF_ACC=1 go test ./aws -v -count 1 -parallel 10 -run=TestAccAWSEksClusterDataSource_basic
=== RUN   TestAccAWSEksClusterDataSource_basic
=== PAUSE TestAccAWSEksClusterDataSource_basic
=== CONT  TestAccAWSEksClusterDataSource_basic
    data_source_aws_eks_cluster_test.go:18: Step 1/1 error: terraform failed: exit status 1

        stderr:

        Error: error reading EKS Cluster (tf-acc-test-gs09q): ResourceNotFoundException: No cluster found for name: tf-acc-test-gs09q.
        {
          RespMetadata: {
            StatusCode: 404,
            RequestID: "91f46140-c965-4387-bddf-9834ad7beb69"
          },
          Message_: "No cluster found for name: tf-acc-test-gs09q."
        }


--- FAIL: TestAccAWSEksClusterDataSource_basic (5.87s)
FAIL
FAIL    github.com/terraform-providers/terraform-provider-aws/aws       5.909s
FAIL
```

Comment (anGie44) [reply aws/resource_aws_eks_cluster.go:118]: Ahh which version of Terraform do you have locally? I've come across these type of errors when using the family of `0.13` versions. I'd recommend`v.0.12.29` or `0.14+`

Comment (philnichol) [reply aws/resource_aws_eks_cluster.go:118]: 🤦 sorry I didn't try that before, works now, will aim to address this and the documentation comment shortly, thanks @anGie44 !

Comment (anGie44) [reply aws/resource_aws_eks_cluster.go:118]: no worries - i don't think we have that behavior documented unfortunately ..sounds great 👍

Comment (philnichol) [reply aws/resource_aws_eks_cluster.go:118]: @anGie44 I've pushed my changes to the data source, let me know if there are any issues. Acceptance tests ran here:

```
$ TF_ACC=1 go test ./aws -v -count 1 -parallel 10 -run=TestAccAWSEksClusterDataSource_basic -timeout 60m
=== RUN   TestAccAWSEksClusterDataSource_basic
=== PAUSE TestAccAWSEksClusterDataSource_basic
=== CONT  TestAccAWSEksClusterDataSource_basic
--- PASS: TestAccAWSEksClusterDataSource_basic (998.81s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       998.846s
```

File: website/docs/r/eks_cluster.html.markdown
Comment (philnichol) [reply website/docs/r/eks_cluster.html.markdown]: good shout, have done this, let me know if the phrasing is alright.

File: aws/resource_aws_eks_cluster.go
Comment (etoews) [inline aws/resource_aws_eks_cluster.go:132]: Please add the 100.64.0.0/10 range here too as per [How do I use multiple CIDR ranges with Amazon EKS?](https://aws.amazon.com/premiumsupport/knowledge-center/eks-multiple-cidr-ranges/) and [Amazon EKS now supports additional VPC CIDR blocks](https://aws.amazon.com/about-aws/whats-new/2018/10/amazon-eks-now-supports-additional-vpc-cidr-blocks/).

Sounds like the 198.19.0.0/16 range is valid in this scenario too but I'm not familiar with that and wouldn't want any ambiguity around it holding up this PR.

File: aws/resource_aws_eks_cluster.go
Comment (anGie44) [reply aws/resource_aws_eks_cluster.go:118]: thanks @philnichol -- test looks good on my end as well 👍 i'm running the resource acceptance tests as well in the background (I don't expect any changes)..and will comment back w/the PR approval

File: website/docs/r/eks_cluster.html.markdown
Comment (etoews) [inline website/docs/r/eks_cluster.html.markdown:190]: You'd need to doc the 100.64.0.0/10 range (and 198.19.0.0/16 range if you want to add it) here too.

**Review by etoews — CHANGES_REQUESTED** [review]

File: aws/resource_aws_eks_cluster.go
Comment (anGie44) [reply aws/resource_aws_eks_cluster.go:132]: thanks for bringing this up @etoews! looks like even though the doc suggests it for use in associating the range to a VPC, we're unfortunately blocked by the AWS API's RFC1918 validation e.g. i gave it a shot locally (removing the validation funcs in place for testing) and passing the CIDR `100.64.0.0/16` (since the netmask size must be between 12 and 24 inclusive):
```
 Error: error creating EKS Cluster (tf-acc-test-blljj): InvalidParameterException: Invalid Cluster Service Ipv4 CIDR for the cluster tf-acc-test-blljj, must be within RFC1918 range
        {
          RespMetadata: {
            StatusCode: 400,
            RequestID: "dcc64171-1ffa-4388-a1e1-e05f4d6cb1d0"
          },
          ClusterName: "tf-acc-test-blljj",
          Message_: "Invalid Cluster Service Ipv4 CIDR for the cluster tf-acc-test-blljj, must be within RFC1918 range"
        }
```

Comment (anGie44) [reply aws/resource_aws_eks_cluster.go:132]: and similar response when passing in `198.19.0.0/16`

Comment (etoews) [reply aws/resource_aws_eks_cluster.go:132]: Yep. I was thinking about another use for those ranges. My misunderstanding. The 100.64.0.0/10 and 198.19.0.0/16 ranges are for pod IP ranges, whereas this PR is for service IP ranges. Apologies for the noise!

Comment (anGie44) [reply aws/resource_aws_eks_cluster.go:132]: no worries at all @etoews!

**Review by anGie44 — APPROVED** [review]
Thanks again for this feature @philnichol 🚀 
Output of acceptance tests:
```
--- PASS: TestAccAWSEksClusterAuthDataSource_basic (17.33s)
--- PASS: TestAccAWSEksCluster_EncryptionConfig (859.07s)
--- PASS: TestAccAWSEksCluster_Logging (895.26s)
--- PASS: TestAccAWSEksCluster_basic (919.82s)
--- PASS: TestAccAWSEksCluster_Tags (948.84s)
--- PASS: TestAccAWSEksClusterDataSource_basic (969.99s)
--- PASS: TestAccAWSEksCluster_VpcConfig_SecurityGroupIds (999.36s)
--- PASS: TestAccAWSEksCluster_VpcConfig_PublicAccessCidrs (1377.22s)
--- PASS: TestAccAWSEksCluster_VpcConfig_EndpointPublicAccess (1632.31s)
--- PASS: TestAccAWSEksCluster_NetworkConfig_ServiceIpv4Cidr (2127.25s)
--- PASS: TestAccAWSEksCluster_VpcConfig_EndpointPrivateAccess (2502.05s)
--- PASS: TestAccAWSEksCluster_Version (2609.50s)
```

**Pull request merged**
